components/plurals/src/rules/runtime/ast.rs

```diff
@@ -0,0 +1,634 @@
+}
+
+impl FromStr for Rule<'_> {
+    type Err = reference::parser::ParserError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let rule = reference::parser::parse(s.as_bytes())?;
+        Ok((&rule).into())
+    }
+}
+
+impl fmt::Display for Rule<'_> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let reference = reference::ast::Rule::from(self);
+        reference::serialize(&reference, f)
+    }
+}
+
+/// `RelationULE` is a type optimized for efficent storing and
+/// deserialization of `rule::runtime::ast::Rule` `ZeroVec` model.
+///
+/// The serialization model packages the rule into 4 bytes plus
+/// variable length of the list of ranges.
+///
+/// The first byte is used to store:
+///  * One bit to store the state of AndOr.
+///  * One bit to store the state of Polarity.
+///  * Six bits to store the Operand.
+///
+/// The following four bytes store the value of the modulo as `u32`.
+///
+/// Finally, the remaining variable length is used to store a list
+/// of `RangeOrValue` elements which have `[u32;2]` layout.
+///
+/// # Diagram
+///
+/// ```text
+/// ┌───────────────┬───────────────┬───────────────┐
+/// │      u8       │    [u8;4]     │     [u8]      │
+/// ├─┬─┬─┬─┬─┬─┬─┬─┼───┬───┬───┬───┼───────────────┤
+/// ├─┼─┼─┴─┴─┴─┴─┴─┼───┴───┴───┴───┼───────────────┤
+/// │ │ │  Operand  │    Modulo     │  RangeOrValue │
+/// └─┴─┴───────────┴───────────────┴───────────────┘
+///  ▲ ▲
+///  │ │
+///  │ Polarity
+///  │
+///  AndOr
+/// ```
+///
+/// # Optimization
+///
+/// This model is optimized for efficient packaging of the `Relation` elements
+/// and performant deserialization from the `RelationULE` to `Relation` type.
+///
+/// # Constraints
+///
+/// The model constraints the `Operand` to 64 variants, and `Modulo` to `u32::MAX`.
+#[derive(Debug)]
+#[repr(packed)]
+pub struct RelationULE {
+    /// This maps to (AndOr, Polarity, Operand),
+    /// with the first bit mapping to AndOr (1 == And), the second bit
+    /// to Polarity (1 == Positive), and the remaining bits to Operand
+    /// encoded via Operand::encode. It is unsound for the Operand bits to
+    /// not be a valid encoded Operand.
+    andor_polarity_operand: u8,
+    modulo: <u32 as AsULE>::ULE,
+    range_list: [RangeOrValueULE],
+}
+
+impl RelationULE {
+    pub fn as_relation(&self) -> Relation {
+        // safe to call because we are operating on a valid RelationULE
+        let (and_or, polarity, operand) =
+            unsafe { Self::decode_andor_polarity_operand(self.andor_polarity_operand) };
+        Relation {
+            and_or,
+            polarity,
+            operand,
+            modulo: u32::from_unaligned(self.modulo),
+            range_list: ZeroVec::borrowed_from_slice(&self.range_list),
+        }
+    }
+
+    fn encode_andor_polarity_operand(and_or: AndOr, polarity: Polarity, operand: Operand) -> u8 {
+        // XXX: Ensure and_or is one bit, polarity is one bit, and operand is max 6 bits
```

Comment:
and_or and polarity are guaranteed 1 bit because they are booleans.  It can't hurt to add a debug_assert for `operand.idx()`.